scripts/bert/create_pretraining_data.py

```diff
@@ -0,0 +1,464 @@
+
+    def __init__(self, tokens, segment_ids, masked_lm_positions,
+                 masked_lm_labels, is_random_next):
+        self.tokens = tokens
+        self.segment_ids = segment_ids
+        self.is_random_next = is_random_next
+        self.masked_lm_positions = masked_lm_positions
+        self.masked_lm_labels = masked_lm_labels
+
+    def __str__(self):
+        s = ''
+        s += 'tokens: %s\n' % (' '.join(
+            [tokenization.printable_text(x) for x in self.tokens]))
+        s += 'segment_ids: %s\n' % (' '.join(
+            [str(x) for x in self.segment_ids]))
+        s += 'is_random_next: %s\n' % self.is_random_next
+        s += 'masked_lm_positions: %s\n' % (' '.join(
+            [str(x) for x in self.masked_lm_positions]))
+        s += 'masked_lm_labels: %s\n' % (' '.join(
+            [tokenization.printable_text(x) for x in self.masked_lm_labels]))
+        s += '\n'
+        return s
+
+    def __repr__(self):
+        return self.__str__()
+
+
+def write_instance_to_example_files(instances, tokenizer, max_seq_length,
+                                    max_predictions_per_seq, output_files):
+    """Create RecordIO files from `TrainingInstance`s."""
+    writers = []
+    for output_file in output_files:
+        writers.append(mx.recordio.MXRecordIO(output_file, 'w'))
+
+    writer_index = 0
+
+    total_written = 0
+    for (inst_index, instance) in enumerate(instances):
+        input_ids = tokenizer.convert_tokens_to_ids(instance.tokens)
+        input_mask = [1] * len(input_ids)
+        segment_ids = list(instance.segment_ids)
+        assert len(input_ids) <= max_seq_length
+
+        while len(input_ids) < max_seq_length:
+            input_ids.append(0)
+            input_mask.append(0)
+            segment_ids.append(0)
+
+        assert len(input_ids) == max_seq_length
+        assert len(input_mask) == max_seq_length
+        assert len(segment_ids) == max_seq_length
+
+        masked_lm_positions = list(instance.masked_lm_positions)
+        masked_lm_ids = tokenizer.convert_tokens_to_ids(
+            instance.masked_lm_labels)
+        masked_lm_weights = [1.0] * len(masked_lm_ids)
+
+        while len(masked_lm_positions) < max_predictions_per_seq:
+            masked_lm_positions.append(0)
+            masked_lm_ids.append(0)
+            masked_lm_weights.append(0.0)
+
+        next_sentence_label = 1 if instance.is_random_next else 0
+
+        features = collections.OrderedDict()
+        features['input_ids'] = input_ids
+        features['input_mask'] = input_mask
+        features['segment_ids'] = segment_ids
+        features['masked_lm_positions'] = masked_lm_positions
+        features['masked_lm_ids'] = masked_lm_ids
+        features['masked_lm_weights'] = masked_lm_weights
+        features['next_sentence_labels'] = [next_sentence_label]
+
+        example = features
+
+        writers[writer_index].write(pickle.dumps(example))
```

Comment:
It might be inefficient to store a RecordIO file of pickled entries? Pickling is very expensive, and is now applied to every sentence individuall during save & load.

If the data fits in memory, consider using a numpy array for serialization. It should allow speeding things up by more than an order of magnitude. When dealing with variable length data, you must concatenate the variable length arrays to a single array and remember the concatenation-points so that you can break up the array again when loading it. See eg https://github.com/dmlc/gluon-nlp/blob/8ec91d8/scripts/machine_translation/dataprocessor.py#L33-L65

What do you think?

Reply:
(Regarding the code I linked above, it would be simpler to use `numpy.split` instead of the list comprehension and use of `zip` in the load function.)

Reply:
Thanks for comments . Also, there is several performance issues to store the duplicated keys for every records.

Reply:
Possibly https://github.com/dmlc/gluon-nlp/pull/498 can be extended (or similar) to take care of and hide the np.concatenate & np.split